File: algorithms/medium/palindrome_partitioning.ex
defmodule Solution do
  use Agent

  @spec partition(s :: String.t()) :: [[String.t()]]
  def partition(s) do
    dfs(s, String.length(s), 0, [], [])
  end

  def dfs(_s, len, l, subs, ans) when l == len do
    [Enum.reverse(subs) | ans]
  end

  def dfs(s, len, l, subs, ans) do
    l..(len - 1)
    |> Enum.reduce(ans, fn r, ans ->
      ss = String.slice(s, l..r)

      if ss == String.reverse(ss) do
        dfs(s, len, r + 1, [ss | subs], ans)
      else
        ans
      end
    end)
  end
end

File: algorithms/medium/palindrome_partitioning.cpp
class Solution {
public:
  vector<vector<string>> ans;
  vector<string> subs;
  vector<vector<string>> partition(string s) {
    vector<string> subs;
    dfs(s, 0);
    return ans;
  }
  void dfs(const string &s, int l) {
    if (l == s.size()) {
      ans.push_back(subs);
      return;
    }
    for (int r = l; r < s.size(); r++) {
      if (is_palindrome(s, l, r)) {
        subs.push_back(s.substr(l, r - l + 1));
        dfs(s, r + 1);
        subs.pop_back();
      }
    }
  }

  bool is_palindrome(const string &s, int l, int r) {
    while (l < r) {
      if (s[l++] != s[r--])
        return false;
    }
    return true;
  }
};
